core/arch/x86_code.c

```diff
@@ -508,4 +508,189 @@ unit_test_asm(dcontext_t *dc)
     test_call_switch_stack(dc);
     test_cpuid();
 }
+
+#    ifdef __AVX__
+
+static void
+write_ymm_aux(dr_ymm_t *buffer, int regno)
+{
+    switch (regno) {
+#        define MOVE_TO_YMM_VEX(buf, num) \
+            asm volatile("vmovdqu %0, "   \
+                         "%%ymm" #num     \
+                         :                \
+                         : "m"(buf[num]))
+#        define MOVE_TO_YMM_EVEX(buf, num) \
+            asm volatile("vmovdqu32 %0, "  \
+                         "%%ymm" #num      \
+                         :                 \
+                         : "m"(buf[num]))
+#        define CASE_MOVE_TO_YMM_VEX(buf, num) \
+        case num:                              \
+            MOVE_TO_YMM_VEX(buf, num);         \
+            break;
+#        define CASE_MOVE_TO_YMM_EVEX(buf, num) \
+        case num: MOVE_TO_YMM_EVEX(buf, num); break;
+
+        CASE_MOVE_TO_YMM_VEX(buffer, 0)
+        CASE_MOVE_TO_YMM_VEX(buffer, 1)
+        CASE_MOVE_TO_YMM_VEX(buffer, 2)
+        CASE_MOVE_TO_YMM_VEX(buffer, 3)
+        CASE_MOVE_TO_YMM_VEX(buffer, 4)
+        CASE_MOVE_TO_YMM_VEX(buffer, 5)
+        CASE_MOVE_TO_YMM_VEX(buffer, 6)
+        CASE_MOVE_TO_YMM_VEX(buffer, 7)
+#        ifdef X64
+        CASE_MOVE_TO_YMM_VEX(buffer, 8)
+        CASE_MOVE_TO_YMM_VEX(buffer, 9)
+        CASE_MOVE_TO_YMM_VEX(buffer, 10)
+        CASE_MOVE_TO_YMM_VEX(buffer, 11)
+        CASE_MOVE_TO_YMM_VEX(buffer, 12)
+        CASE_MOVE_TO_YMM_VEX(buffer, 13)
+        CASE_MOVE_TO_YMM_VEX(buffer, 14)
+        CASE_MOVE_TO_YMM_VEX(buffer, 15)
+#            ifdef __AVX512F__
+        CASE_MOVE_TO_YMM_EVEX(buffer, 16)
+        CASE_MOVE_TO_YMM_EVEX(buffer, 17)
+        CASE_MOVE_TO_YMM_EVEX(buffer, 18)
+        CASE_MOVE_TO_YMM_EVEX(buffer, 19)
+        CASE_MOVE_TO_YMM_EVEX(buffer, 20)
+        CASE_MOVE_TO_YMM_EVEX(buffer, 21)
+        CASE_MOVE_TO_YMM_EVEX(buffer, 22)
+        CASE_MOVE_TO_YMM_EVEX(buffer, 23)
+        CASE_MOVE_TO_YMM_EVEX(buffer, 24)
+        CASE_MOVE_TO_YMM_EVEX(buffer, 25)
+        CASE_MOVE_TO_YMM_EVEX(buffer, 26)
+        CASE_MOVE_TO_YMM_EVEX(buffer, 27)
+        CASE_MOVE_TO_YMM_EVEX(buffer, 28)
+        CASE_MOVE_TO_YMM_EVEX(buffer, 29)
+        CASE_MOVE_TO_YMM_EVEX(buffer, 30)
+        CASE_MOVE_TO_YMM_EVEX(buffer, 31)
+#            endif
+#        endif
+    default: FAIL(); break;
+    }
+}
+
+void
```

Comment:
static, and called by unit_test_asm(), which is the top-level driver here.  Ditto below.

Reply:
Done.